var/spack/repos/builtin/packages/llvm/package.py

```diff
@@ -581,12 +589,13 @@ def cmake_args(self):
     @run_before("build")
     def pre_install(self):
         with working_dir(self.build_directory):
-            # When building shared libraries these need to be installed first
-            make("install-LLVMTableGen")
-            if self.spec.version >= Version("4.0.0"):
-                # LLVMDemangle target was added in 4.0.0
-                make("install-LLVMDemangle")
-            make("install-LLVMSupport")
+            if "+shared_libs" in self.spec:
+               # When building shared libraries these need to be installed first
+               make("install-LLVMTableGen")
```

Comment:
So, what was the story with this? Did you try `+shared_libs` with the llvm+ninja build? You mentioned that failed https://github.com/spack/spack/issues/23880#issuecomment-848257616

Reply:
`shared_libs` defaults to `False` so I added the check.

Reply:
Yes, but if you have it enabled, does it build with ninja or not?

Reply:
Hm, maybe this is outside the scope of making ninja work. We should probably replace `BUILD_SHARED_LIBS` with `LLVM_BUILD_LLVM_DYLIB` as recommended by the llvm docs.

Reply:
> Yes, but if you have it enabled, does it build with ninja or not?

Clearly not.

Reply:
Now I am wondering that pre build phase manual `make` commands are required or not.

Reply:
For `llvm` switching to `Ninja` only is a sensible option as it is much faster and the community uses it.

Reply:
But I'd like to hear @naromero77 and @trws 's opinions on this. What's your take?

Reply:
```
       if "+shared_libs" in spec:
            cmake_args.append("-DBUILD_SHARED_LIBS:Bool=ON")
        if "+llvm_dylib" in spec:
            cmake_args.append("-DLLVM_BUILD_LLVM_DYLIB:Bool=ON")
```

I'm stumbled upon these lines.

Reply:
Yeah, I noticed these too. I would pressume clang is not really usable with `BUILD_SHARED_LIBS=On`. Should we consider dropping that option @naromero77 / @trws? "Advanced" users can just run `spack build-env llvm+clang` and then configure with -DBUILD_SHARED_LIBS=On by hand?

Reply:
> Now I am wondering that pre build phase manual `make` commands are required or not.

@trws would know more than me

Reply:
> For `llvm` switching to `Ninja` only is a sensible option as it is much faster and the community uses it.

I think Ninja-only is a sensible option. But see my comments later on in this review.

Reply:
Ninja only seems fine, but anything in there that used to build with the make-based version should build with the ninja-based version.  Replacing those `which`-based make lines with something from `which('ninja')` should just work, but needs to be tested.

Reply:
Alternately, it could be re-worked to use cmake instead to make it portable to keep both with something like `cmake("--build", ".", "--target", "install-LLVMTableGen")`.